entity-framework/core/what-is-new/ef-core-9.0/whatsnew.md

```diff
@@ -635,6 +637,59 @@ FROM [Posts] AS [p]
+
+<!--
+#region DefaultParameterizationPrimitiveCollection
+async Task<List<Post>> GetPostsPrimitiveCollection(int[] ids)
+    => await context.Posts
+        .Where(e => e.Title == ".NET Blog" && ids.Contains(e.Id))
+        .ToListAsync();
+-->
+[!code-csharp[ForceParameter](../../../../samples/core/Miscellaneous/NewInEFCore9/QuerySample.cs?name=DefaultParameterizationPrimitiveCollection)]
+
+will result in the following translation on SQL Server:
+
+```output
+Executed DbCommand (5ms) [Parameters=[@__ids_0='[1,2,3]' (Size = 4000)], CommandType='Text', CommandTimeout='30']
+SELECT [p].[Id], [p].[Archived], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Discriminator], [p].[PublishedOn], [p].[Rating], [p].[Title], [p].[PromoText], [p].[Metadata]
+FROM [Posts] AS [p]
+WHERE [p].[Title] = N'.NET Blog' AND [p].[Id] IN (
+    SELECT [i].[value]
+    FROM OPENJSON(@__ids_0) WITH ([value] int '$') AS [i]
+)
+```
+
+This allows having the same SQL query for different parameterized collections (only the parameter value changes), but in some situations it can lead to performance issues as the database isn't able to optimally plan for the query. `EF.Constant` method can be used to revert to the previous translation.
+
+The following query uses `EF.Constant` to that effect:
+
+<!--
+#region ForceConstantPrimitiveCollection
+async Task<List<Post>> GetPostsForceConstantCollection(int[] ids)
+    => await context.Posts
+        .Where(e => e.Title == ".NET Blog" && EF.Constant(ids).Contains(e.Id))
+        .ToListAsync();
+-->
+[!code-csharp[ForceParameter](../../../../samples/core/Miscellaneous/NewInEFCore9/QuerySample.cs?name=ForceConstantPrimitiveCollection)]
+
+The resulting SQL is as follows:
+
+```sql
+SELECT [p].[Id], [p].[Archived], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Discriminator], [p].[PublishedOn], [p].[Rating], [p].[Title], [p].[PromoText], [p].[Metadata]
+FROM [Posts] AS [p]
+WHERE [p].[Title] = N'.NET Blog' AND [p].[Id] IN (1, 2, 3)
+```
+
+Moreover, EF9 introduces `TranslateParameterizedCollectionsToConstants` [context option](/ef/core/dbcontext-configuration/#dbcontextoptions) that can be used to prevent primitive collection parameterization for all queries. We also added a complementing `TranslateParameterizedCollectionsToParameters` which forces parameterization of primitive collections explicitly (this is the default behavior).
+
+> [!TIP]
+> `EF.Parameter` method overrides the context option. If you want to prevent parameterization of primitive collections for most of your queries (but not all), you can set the context option `TranslateParameterizedCollectionsToConstants` and use `EF.Parameter` for the queries or individual variables that you want to parameterize.
```

Comment:
```suggestion
> The `EF.Parameter` method overrides the context option. If you want to prevent parameterization of primitive collections for most of your queries (but not all), you can set the context option `TranslateParameterizedCollectionsToConstants` and use `EF.Parameter` for the queries or individual variables that you want to parameterize.
```